### PR TITLE
Move TypeLoadException to shared

### PR DESCRIPTION
**Pull request by EgorBo**
Or it's better to if-def CoreCLR specific code for changes like this?

### CODE REVIEW

File: src/System.Private.CoreLib/src/System/TypeLoadException.CoreCLR.cs [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+**
+**
+=============================================================================*/
+
+using System.Runtime.InteropServices;
+using System.Runtime.CompilerServices;
+
+namespace System
+{
+    [System.Runtime.CompilerServices.TypeForwardedFrom("mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089")]
```
Comment (jkotas) [inline src/System.Private.CoreLib/src/System/TypeLoadException.CoreCLR.cs]: This should be in the share part.

**Comment by jkotas** [comment]
Fix the field names to follow the coding conventions?

**Comment by jkotas** [comment]
Can `ResourceId` be private?

**Comment by jkotas** [comment]
> Or it's better to if-def CoreCLR specific code for changes like this?

How much of the the logic moved to shared is Mono going to take advantage of?

File: src/System.Private.CoreLib/src/System/TypeLoadException.CoreCLR.cs (hunk 1 above)
Comment (EgorBo) [reply src/System.Private.CoreLib/src/System/TypeLoadException.CoreCLR.cs]: Both Mono and CoreRT don't have it, should I still move the attribute there?

Comment (jkotas) [reply src/System.Private.CoreLib/src/System/TypeLoadException.CoreCLR.cs]: CoreRT has it: https://github.com/dotnet/corert/blob/master/src/System.Private.CoreLib/src/System/TypeLoadException.cs#L10

And all exceptions in shared have it too.

Comment (EgorBo) [reply src/System.Private.CoreLib/src/System/TypeLoadException.CoreCLR.cs]: ah indeed

**Comment by EgorBo** [comment]
> > Or it's better to if-def CoreCLR specific code for changes like this?
> 
> How much of the the logic moved to shared is Mono going to take advantage of?

um.. not sure I understand the question, the whole file except the qcall

BTW, addressed the feedback

**Comment by jkotas** [comment]
> not sure I understand the question

E.g. does Mono use and set the `_resourceId` field?

**Comment by EgorBo** [comment]
@jkotas the old mono did used `ResourceId` property, but I guess since coreclr/corefx don't use it anywhere we also don't need it, @marek-safar thoughts?

File: src/System.Private.CoreLib/shared/System/TypeLoadException.cs [hunk 2]
```diff
@@ -106,30 +68,27 @@ public string TypeName
...
         {
-            ClassName = info.GetString("TypeLoadClassName");
-            AssemblyName = info.GetString("TypeLoadAssemblyName");
-            MessageArg = info.GetString("TypeLoadMessageArg");
-            ResourceId = info.GetInt32("TypeLoadResourceID");
+            _className = info.GetString("TypeLoadClassName");
+            _assemblyName = info.GetString("TypeLoadAssemblyName");
+            _messageArg = info.GetString("TypeLoadMessageArg");
+            _resourceId = info.GetInt32("TypeLoadResourceID");
         }
```
Comment (marek-safar) [inline src/System.Private.CoreLib/shared/System/TypeLoadException.cs]: private TypeLoadException ctor is CoreCLR specific

**Comment by marek-safar** [comment]
> E.g. does Mono use and set the _resourceId field?

it does not but I expect we need to serialize it for compatibility reasons. Also not sure what the globalization story will be going forward so we might need to switch to using it.

File: src/System.Private.CoreLib/shared/System/TypeLoadException.cs [hunk 3]
```diff
@@ -0,0 +1,76 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+/*=============================================================================
+**
+**
+**
+** Purpose: The exception class for type loading failures.
```
Comment (jkotas) [inline src/System.Private.CoreLib/shared/System/TypeLoadException.cs]: Nit: We have been deleting useless comments like this when moving stuff to shared.

**Pull request merged**